### ISSUE TITLE
annexscientificname link to scientificname

### ISSUE THREAD

**Issue by damianooldoni** [inbo/speciesbim#36]
We decided to link scientificnameannex to scientificname (not to taxonomy). Names with a scientificname_id are easy to match to the taxonomy (over the scientificname table). Those without a scientificname_id, such as informal names are more complex descriptions. To link ALL names for an annex to the taxonomy, BIM will have to make bespoke views combining directly linkable names with queries for certain groups.

**Comment by damianooldoni** [comment]
Based on discussion with @peterdesmet  and @niconoe we think it's better to change the name of the table `scientificnameannex` to `annexscientificname`.

**Comment by damianooldoni** [comment]
While working on #36 I found that names in `scientificname` table are not unique. 
As you can see in table below, aside same `scientificName`, we CAN have same `authorhsip` values.

As briefly discussed in chat with @niconoe, we could add `DISTINCT` on `scientificName` and `authorship` in [SQL statement](https://github.com/inbo/speciesbim/blob/master/scripts/sql_snippets/populate_scientificname.sql) to populate `scientificname`.
So, while making the table `scientificname` should we put a constraint about the uniqueness of the combination of these two columns? @niconoe: is it possible?
Still, to take in mind while deciding about this: names in `annexscientificname` table are almost always WITHOUT authorship.

id | deprecatedTaxonId | scientificName | authorhsip
-- | -- | -- | --
4859 | 26718 | Agaricus sylvicola | (Vittad.) Peck, 1872
6818 | 50129 | Agaricus sylvicola | 
7743 | 51280 | Agonum emarginatum | (Gyllenhal, 1827)
2169 | 16405 | Agonum emarginatum | (Gyllenhal, 1827)
4871 | 26780 | Amanita excelsa | (Fr.:Fr.) Bertill.
7945 | 51486 | Amanita excelsa | Bertill., 1866 (Fr.)
8252 | 51819 | Amanita excelsa | Bertill., 1866 (Fr.)
8138 | 51681 | Anas clypeata | Linnaeus, 1758
720 | 3521 | Anas clypeata | Linnaeus, 1758
713 | 3510 | Anas penelope | Linnaeus, 1758
8146 | 51689 | Anas penelope | Linnaeus, 1758
8156 | 51699 | Anas querquedula | Linnaeus, 1758
715 | 3512 | Anas querquedula | Linnaeus, 1758
8140 | 51683 | Anas sibilatrix | Poeppig, 1829
6842 | 50153 | Anas sibilatrix | Poeppig, 1829
716 | 3513 | Anas strepera | Linnaeus, 1758
8139 | 51682 | Anas strepera | Linnaeus, 1758
2956 | 20772 | Anchusa arvensis | (L.) M.Bieb.
8012 | 51554 | Anchusa arvensis | M. Bieb. (L.)
2527 | 18127 | Anoplotrupes stercorosus | (Scriba, 1791)
8154 | 51697 | Anoplotrupes stercorosus | (Scriba, 1791)
6874 | 50185 | Arabidopsis arenosa subsp. arenosa | (L.) Lawalrée
6873 | 50184 | Arabidopsis arenosa subsp. arenosa | (L.) Lawalrée
6877 | 50188 | Arctium minus | (Hill) Bernh.
2882 | 20672 | Arctium minus | (Hill) Bernh.
5831 | 34306 | Bombus | Latreille, 1802
7718 | 51039 | Bombus | Latreille, 1802
7888 | 51429 | Bromus madritensis | L.
7961 | 51502 | Bromus madritensis | L.
6944 | 50252 | Calamagrostis epigejos | (L.) Roth
3301 | 21247 | Calamagrostis epigejos | (L.) Roth
8285 | 51853 | Caloplaca aurantia | Hellb. (Pers.)
5796 | 34158 | Caloplaca aurantia | (Pers.) Hellbom
8288 | 51856 | Caloplaca decipiens | Blomb. & Forssell (Arnold)
5799 | 34165 | Caloplaca decipiens | (Arnold) Blomb. & Forssell
8293 | 51861 | Caloplaca flavescens | J.R.Laundon (Huds.)
5800 | 34166 | Caloplaca flavescens | (Hudson) J.R. Laundon
5802 | 34169 | Caloplaca flavovirescens | (Wulfen) Dalla Torre & Sarnth.
8304 | 51872 | Caloplaca flavovirescens | Dalla Torre & Sarnth. (Wulfen)
8074 | 51616 | Cardamine flexuosa | With.
2979 | 20801 | Cardamine flexuosa | With.
7663 | 50980 | Carduelis flammea | (Linnaeus, 1758)
698 | 3484 | Carduelis flammea | (Linnaeus, 1758)
4949 | 27588 | Clitocybe gibba | (Pers.:Fr.) P. Kumm.
8256 | 51823 | Clitocybe gibba | P.Kumm. (Pers.)
7028 | 50333 | Cortinarius livido-ochraceus | (Berk.) Berk. 1860
986 | 4624 | Cortinarius livido-ochraceus | (Berk.) Berk.
8124 | 51666 | Craspedosoma rawlinsii | Leach, 1815
1233 | 5704 | Craspedosoma rawlinsii | Leach, 1814
3411 | 21410 | Crataegus laevigata | (Poir.) DC.
7773 | 51313 | Crataegus laevigata | DC. (Poir.)
7754 | 51293 | Cryphia domestica | Hufnagel, 1766
4011 | 22818 | Cryphia domestica | (Hüfnagel, 1766)
601 | 3264 | Dactylorhiza fuchsii | (Druce) Soó
7731 | 51054 | Dactylorhiza fuchsii | Soó (Druce)
667 | 3435 | Dendrocopos medius | (Linnaeus, 1758)
8141 | 51684 | Dendrocopos medius | (Linnaeus, 1758)
8142 | 51685 | Dendrocopos minor | (Linnaeus, 1758)
668 | 3436 | Dendrocopos minor | (Linnaeus, 1758)
7079 | 50385 | Draba verna | L.
7785 | 51325 | Draba verna | L.
7089 | 50395 | Enoplognatha | 
6457 | 41327 | Enoplognatha | 
7143 | 50450 | Galerina vittiformis | (Fr.) Singer 1950
8241 | 51784 | Galerina vittiformis | Singer, 1950 (Fr.)
8011 | 51553 | Galium parisiense | L.
7794 | 51334 | Galium parisiense | L.
5177 | 29438 | Hemimycena lactea | (Pers.:Fr.) Singer
8258 | 51825 | Hemimycena lactea | Singer (Pers.)
8242 | 51785 | Hygrocybe conica | P.Kumm. (Schaeff.)
5205 | 29642 | Hygrocybe conica | (Schaeff.:Fr.) P. Kumm.
1048 | 4988 | Hygrocybe glutinipes | (J.E. Lange) R. Haller
8118 | 51660 | Hygrocybe glutinipes | R. Haller Aar., 1956 (J.E. Lange)
3070 | 20922 | Hypericum dubium | Leers
8109 | 51651 | Hypericum dubium | Leers
7957 | 51498 | Lactarius vellereus | Fr., 1838 (Fr.)
7248 | 50556 | Lactarius vellereus | (Fr.) Fr. 1838
2995 | 20822 | Lepidium draba | L.
8082 | 51624 | Lepidium draba | L.
8286 | 51854 | Lepraria lobificans | Nyl.
7269 | 50577 | Lepraria lobificans | Nyl.
8254 | 51821 | Lycoperdon pyriforme | Schaeff.
5298 | 30160 | Lycoperdon pyriforme | Schaeff.:Pers.
7736 | 51059 | Malus sylvestris subsp. sylvestris | 
7301 | 50609 | Malus sylvestris subsp. sylvestris | 
875 | 4246 | Morchella esculenta | (L.:Fr.) Pers.
8151 | 51694 | Morchella esculenta | Pers. (L.)
5337 | 30529 | Myxarium nucleatum | Wallr.
7900 | 51441 | Myxarium nucleatum | Wallr., 1833
7056 | 50676 | Oulema | 
6039 | 37486 | Oulema | Goz., 1886
5400 | 30875 | Panellus serotinus | (Pers.) Kühner
7935 | 51476 | Panellus serotinus | Kühner, 1950 (Pers.)
889 | 4269 | Phallus impudicus | L.:Pers.
7824 | 51365 | Phallus impudicus | L., 1753
7426 | 50737 | Phyllonorycter klemannella | (Fabricius, 1781)
3877 | 22507 | Phyllonorycter klemannella | (Fabricius, 1781)
4445 | 24425 | Polistes dominula | (Christ, 1791)
7461 | 50771 | Polistes dominula | (Christ, 1791)
7979 | 51520 | Pseudepidalea viridis | (Laurenti, 1768)
8251 | 51794 | Pseudepidalea viridis | (Laurenti, 1768)
7860 | 51401 | Securigera varia | Lassen (L.)
3145 | 21036 | Securigera varia | (L.) Lassen
3506 | 21519 | Solanum triflorum | Nutt.
7899 | 51440 | Solanum triflorum | Nutt.
7734 | 51057 | Stellaria media subsp. media | 
7571 | 50881 | Stellaria media subsp. media | (L.) Vill.
7928 | 51469 | Stigmatogaster subterranea | (Shaw, 1789)
1254 | 5762 | Stigmatogaster subterranea | (Shaw, 1789)
7849 | 51390 | Tubaria furfuracea | Gillet, 1876 (Pers.)
7627 | 50937 | Tubaria furfuracea | (Pers.) Gillet 1876
8246 | 51789 | Ulmus minor | Mill.
3515 | 21531 | Ulmus minor | Mill.
8292 | 51860 | Xanthoria elegans | Th.Fr. (Link)
5811 | 34181 | Xanthoria elegans | (Link) Th. Fr.

**Comment by niconoe** [comment]
Yep, technically speaking there's no issue of setting a constraint of a combination of two columns, such as `scientificName` and `authorship`.

Content-wise, I think we have two different cases:

- Entries with both the same scientific name and the same authors (such as `Ulmus minor`) shouldn't happen at all IMHO, this can be considered as a bug in our scripts (that we can solve defensively, by adding a constraint first then solving then script)
- For the cases where we have the same scientific names with different authors (because they appear like that in other parts of the BIM database), I think it's BIM responsibility to tell us what they need: the two obvious (but imperfect) solutions would be 1) to keep duplicates (current situation) or 2) to reconcile with existing data (no duplicates, but only a single author string for each scientificname). In that case, which author string should take precedence over the others? @billobah, any opinion about that?

**Comment by damianooldoni** [comment]
The fact that the number of duplicates is relatively small (57 distinct scientific names) makes option 2 possible. But, agree with @niconoe: it's up to BIM to tell us what it's best for them. I link this discussion to email for Mathias and Olivier as solving this issue involves not only IT staff.

**Comment by damianooldoni** [comment]
Based on internal discussion with @peterdesmet and @niconoe, we think we can solve these duplicates as their solution is quite straightforward, so winning time. As solving these duplicates is something specific and not really involving annexes, we move the discussion about it to this issue: #41.

This issue can still be left open until the link to `scientificname` is done.

**Comment by damianooldoni** [comment]
Some names in official_annexes.csv contain authorship information. So, this should be splitted from the canonical name in the csv so that it is correctly imported and correctly matched against the columns `scientificName` and `authorship` of scientificname table. See example of some taxa below

scientific_name_original | scientific_name_corrected | authorship
-- | -- | --
Asplenium jahandiezii (Litard.) Rouy | Asplenium jahandiezii | (Litard.) Rouy
Woodwardia radicans (L.) Sm. | Woodwardia radicans | (L.) Sm.
Culcita macrocarpa C. Presl | Culcita macrocarpa | C. Presl
Diplazium sibiricum (Turez. ex Kunze) Kurata | Diplazium sibiricum | (Turez. ex Kunze) Kurata
Dryopteris corleyi Frase | Dryopteris corleyi | Frase

**Comment by damianooldoni** [comment]
I am implementing the link of annexscientificname to scientificname by using a scientificnameId in table annexscientificname as shown below:

id | scientificnameId | scientificNameOriginal | scientificName | authorship | remarks | annexCode
-- | -- | -- | -- | -- | -- | --
1 | 51 | Barbastella barbastellus | Barbastella barbastellus |  |   | BXL-ORD-2012_Annex II.1
2 | 52 | Myotis bechsteini | Myotis bechsteini |  |   | BXL-ORD-2012_Annex II.1
3 | 53 | Myotis dasycneme | Myotis dasycneme |  |   | BXL-ORD-2012_Annex II.1
4 | 54 | Myotis emarginatus | Myotis emarginatus |  |   | BXL-ORD-2012_Annex II.1
5 | 55 | Myotis Myotis | Myotis myotis  |  | decapitalize myotis | BXL-ORD-2012_Annex II.1
6 | 56 | Rhinolophus ferrumequinum | Rhinolophus ferrumequinum |  |   | BXL-ORD-2012_Annex II.1
7 | 57 | Rhodeus sericeus amarus | Rhodeus sericeus amarus |   |  | BXL-ORD-2012_Annex II.1
18 |  | Alle europese soorten Mammalia |   |  |   | BXL-ORD-2012_Annex II.2

We agree that the names in `scientificname` table can be improved e.g. due to unmatch or mismatch with GBIF backbone. But this will leave the changes in `scientificname` table and not change them back in `annexscientificname` table as there is no link from `scientificname` to `annexscientificname` (a kind of `annexscientificnameId`). Shouldn't be better to update the values in `scientificName` and `authorship` of annexscientificname table as well?

@niconoe, @peterdesmet : what do you think about this? It's more a maintenance problem but if has consequences on the table structure of `scientificname` and so should be tackled now.

**Comment by niconoe** [comment]
@damianooldoni: not sure I have the complete view over this, but:

**Question 1**: couldn't we imagine following the reverse relationship: when an entry is updated in `scientificname`, look for entries in `annexscientificname` that link to it, and also update it? (If so, we could probably make that automatic using a database trigger)

In that case I have a **second question**: should the `scientificname` (the corrected field) in `annexscientificname` ideally be always identical to the scientific name in the `scientificName` table? Or do we have scenario where they should have different values? (in the first case, we might want to refine the table structure further)

**Comment by damianooldoni** [comment]
Question 1: YES

Question 2: YES, the corrected fields `scientificName` and `authorship` in `annexscientificname` table should be ideally always be identical to the same fields in `scientificname` table. But if they have to be the same, maybe we could also drop these two fields from `annexscientificname` and just maintain a two direction relationship. When we populate `annexscientificname`with the original names we add the "corrected" versions directly in `scientificname` table instead of adding them in `annexscientificname`. Would it be not the easier and most robust approach? If they want to see how annex original names are improved then we can do it within a view 

`annexscientificname` table:

id | scientificnameId | scientificNameOriginal | remarks | annexCode
-- | -- | -- | -- | --
14 | 45 | Asbestophluma hypogea | removed h | BXL-ORD-2012_Annex II.2


`scientificname` table:

id | taxonomyId | annexscientificnameId | deprecatedTaxonId | scientificName | authorship | lastMatched | matchConfidence | matchType
-- | -- | -- | -- | -- | -- | -- | -- | --
45 |  | 14 |  | Asbestopluma hypogea |  | | 

This means adapting `populate_annex_scientificname()` function but I think it will be much less work than other solutions. Curious what @niconoe and @peterdesmet think about this!

**Comment by peterdesmet** [comment]
Absolutely agree with last suggestion: `scientificName` and `authorship` should be removed from `annexscientificname`. They should only be created temporally to populate the `scientificname` table during migration. `scientificNameOriginal` can stay: it is the name as it appears on the annex. Maybe that field can be renamed to `scientificNameInAnnex`.

**Comment by niconoe** [comment]
I agree with all this, for the biggest part:

- yes, `scientificName` and `authorship` should disappear from `annexscientificname` (general good practice: no duplicate data in a RDBMS -> leads to inconsistencies)
- I don't think we need to add a reverse FK from `scientificname` to `annexscientificname`, for two reasons: 

1) I don't think it's necessary (a single FK can always be followed backwards, so the relationship can be followed from both sides - you almost never needs such a construct)
2) it would only allow a 1-1 relationship between both tables. But I think we need to have several entries from `annexscientificname` to be able to point to the same entry in `scientificname`, no?

- I also agree with @peterdesmet that `scientificNameInAnnex` is a better (= more explicit) field name.

**Comment by damianooldoni** [comment]
Nice to hear, @niconoe, you agree too.

You are absolutely right, we can have multiple entries in `annexscientificname` table pointing to same name in `scientificname` table! So, both your reasons are valid!

And about the column name choice, it is extremely difficult to find better names and not agree with @peterdesmet. So, even this time I agree on it.

**Comment by damianooldoni** [comment]
Ok. I almost there. 

Based on the last comments and the information in the csv file with annex species, I am going towards this structure:

`annexscientificname` table:

id | scientificnameId | scientificNameInAnnex | remarks | annexCode
-- | -- | -- | -- | --
14 | 45 | Asbestophluma hypogea | removed h | BXL-ORD-2012_Annex II.2 
15 | | Alle Europese soorten Aves | | BXL-ORD-2012_Annex II.2

`scientificname` table:

id | scientificName | authorship
-- | -- | --
45 | Asbestopluma hypogea | 

where the (corrected) names are imported to scientificname table if not already present while expressions (e.g. `Alle Europese soorten Aves`) are not.

This is ok in migration. But in the future, when BIM will add names in `annexscientificname` from new annexes and will run the match to `scientficname` table again, expressions  will be imported to `scientificname` table as well as they don't have a `scientificnameId` yet (and we don't want they have it).

I propose to add a boolean field isExpression (or its counterpart isName) to `annexscientificname` table so that we can select easily which rows should be inserted/matched to `scientificname` table.

@niconoe, @peterdesmet: what do you think about it?

**Comment by peterdesmet** [comment]
Sure, would opt for isScientificName

Also, why not camelCase for scientificnameid? What is the current convention in the database?

I’d also order the fields in the proposed annexscientificname table differently, but maybe it’s consistent with the rest?

**Comment by damianooldoni** [comment]
Thanks @peterdesmet.  I will use `isScientificName`, good idea.

I used `scientificnameId` instead of `scietnficiNameId` because the table is called `scientificname`. But indeed, this is confusing.

About the current convention in the db: BIM table columns are all lowercase. The new ones are camelCase, probably influenced by GBIF? @niconoe: is there any other reason about the use of camelCase? Anyway, I prefer camelCase above all other solutions.

About the order of the fields in the proposed `anenxscientificname` I would prefer to discuss in #46.

**Comment by niconoe** [comment]
Yep, I agree with the need for this field and find `isScientificName` a good name!

To answer your last reason: I don't think there's a specific reason for camelCase. I think we decided that at first, and it's probably worth to review tables at some point to make sure we've always been consistent...

**Comment by peterdesmet** [comment]
We'll then need to review that before we trigger migration. `camelCase` are more readable, but have the disadvantage they they always need to be "quoted" in queries in postgres
